### ISSUE TITLE
no-unmodified-loop-condition false positive

### ISSUE THREAD

**Issue by vegerot** [eslint/eslint#13464]
**Tell us about your environment**

* **ESLint Version:7.3.1**
* **Node Version:14.5**
* **npm Version:6.14.5**

**What parser (default, Babel-ESLint, etc.) are you using?**
`vue-eslint-parser`
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
      for (const d = start; d <= end; d.add(1, 'day')) {
        headers.push({
          align: 'center',
          filterable: false,
          sortable: false,
          text: d.format('ddd DD'),
          value: d.format(CALENDAR_FORMAT),
        });
      }
```

```bash
vue-cli-service lint
```

**What did you expect to happen?**
since I am incrementing `d` after each loop, it is being modified

**What actually happened? Please include the actual, raw output from ESLint.**
```bash
error: 'd' is not modified in this loop (no-unmodified-loop-condition) at src/views/foo.vue:460:29:
  458 |       }
  459 |     } else {
> 460 |       for (const d = start; d <= end; d.add(1, 'day')) {
      |                             ^
  461 |         headers.push({
  462 |           align: 'center',
  463 |           filterable: false,
```

**Are you willing to submit a pull request to fix this bug?**
Sure, with some guidance from the eslint team

**Comment by anikethsaha** [comment]
Thanks for the report. 

IMO, the rule is working fine. 
I think the rule here, checks for the write reference of the variable, but instead, it got, call expression, 
This rule doesn't understand the method `.add()` as it is not in the scope of the rule. so without runtime, it won't be able to know what's happening in the `.add` method. 

It might be solved using typescript and typescript specific rule which will be overriding this rule so that it can assert the properties and do the checking for the return type of the `d.add`.

**Comment by vegerot** [comment]
Interesting.  How could that be done?

**Comment by anikethsaha** [comment]
this might be complex it seems, as return type won't work here. 
what I was thinking is to somehow tell the analyzer that the value is being modified when `.add` is being called. Which I am not aware of how to do so. 
So, 
my suggestion would be to disable the rule fo that line. 
let see if anyone else has some thoughts in this

**Comment by mdjermanovic** [comment]
I agree that this seems to be a limitation of this rule that probably can't be properly fixed.

In the original example, the rule might be able to figure out that `d` is an object, and that it will be converted to a primitive in the loop's condition. But, it wouldn't be always possible, e.g.:

```js
for (const d = start; d <= end; addDay(d)) {
    // ...
}
```

so I'm not sure if we should change the rule, since it would add a lot of complexity and still wouldn't be able to eliminate all false positives.

Maybe `@typescript-eslint/eslint-plugin` could extend the rule and filter out reported non-primitives in a position where they are going to be converted to primitive values, though this might introduce some false negatives.
